### PR TITLE
[CPU] [EXPERIMENTAL] FullyConnected: enabled sparsity weights decompression

### PR DESCRIPTION
**Pull request by antonvor**
Issue: 77198
PR in onednn fork: https://github.com/openvinotoolkit/oneDNN/pull/158

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/core/properties/properties.cpp [hunk 1]
```diff
@@ -71,6 +71,12 @@ void regmodule_properties(py::module m) {
 
     // Submodule intel_cpu property
     wrap_property_RW(m_intel_cpu, ov::intel_cpu::denormals_optimization, "denormals_optimization");
+    wrap_property_RW(m_intel_cpu,
+                     ov::intel_cpu::experimental_fc_sparse_wei_decompression_enable,
+                     "experimental_fc_sparse_wei_decompression_enable");
+    wrap_property_RW(m_intel_cpu,
+                     ov::intel_cpu::experimental_fc_sparse_wei_decompression_rate,
```
Comment (dmitry-gorokhov) [inline src/bindings/python/src/pyopenvino/core/properties/properties.cpp:78]: 1. I would propose to simplify: ov::intel_cpu::sparse_weights_decompression_rate 
2. Since we haven't implemented good heuristic to allow automatic switch for sparse path, lets leave only rate key. 1.0f - default value meaning sparsity is disabled.

File: src/tests/ie_test_utils/common_test_utils/ov_tensor_utils.cpp [hunk 2]
```diff
@@ -234,9 +234,13 @@ void compare(const ov::Tensor& expected,
         auto eps = std::numeric_limits<double>::epsilon();
         return (b - a) >= (std::fmax(std::fabs(a), std::fabs(b)) * eps);
     };
+    // todo: [av] delete
+    // std::cout << "COMPARE" << std::endl;
     for (size_t i = 0; i < shape_size(expected_shape); i++) {
         double expected_value = expected_data[i];
         double actual_value = actual_data[i];
+        // if (std::fabs(expected_value - actual_value) > 0.5)
+        //     std::cout << i << ". expected = " << expected_value << ", actual = " << actual_value << std::endl;
```
Comment (dmitry-gorokhov) [inline src/tests/ie_test_utils/common_test_utils/ov_tensor_utils.cpp:243]: please remove commented code.

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp [hunk 3]
```diff
@@ -668,6 +688,73 @@ InferenceEngine::Precision FullyConnected::getRuntimePrecision() const {
...
+    auto elementsCount = blb->GetDescWithType<BlockedMemoryDesc>()->getPaddedElementsCount();
+    size_t zerosCounts = 0;
+    for (int i = 0; i < elementsCount; i++) {
+        if (weightsData[i] == 0) {
+            zerosCounts++;
+        }
+    }
+    nnzCount = elementsCount - zerosCounts;
+
+    if (printSparseRates) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/fullyconnected.cpp:731]: Shouldn't be part of production solution. You can wrap with DEBUG_LOG macros instead.

File: src/plugins/intel_cpu/src/nodes/fullyconnected.h [hunk 4]
```diff
@@ -75,6 +79,14 @@ class FullyConnected : public Node {
...
     static const size_t BIAS_ID = 2;
     dnnl::memory::data_type outputDataType;
+
+    // sparse weights
+    bool useSparseWeights = false;
+    bool printSparseRates = false; // todo: [av] delete
+    int nnzCount = -1;
+    float minSparseRate = 1.f;
+    float weiSparseRate = 0.f;
+    bool canUseSparseWeights();
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/fullyconnected.h:89]: I would propose: `useSparseWeightsDecompression` name for better alignment with corresponding configuration key.

File: src/plugins/intel_cpu/src/graph.cpp [hunk 5]
```diff
@@ -1444,6 +1448,19 @@ void Graph::EnforceBF16() {
     }
 }
 
+void Graph::EnableSparseWeightsForFC(float minSparseRate) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/graph.cpp:1451]: just `setMinSparseRate` - this is what the function actually do.

File: src/plugins/intel_cpu/src/graph.cpp [hunk 6]
```diff
@@ -1444,6 +1448,19 @@ void Graph::EnforceBF16() {
...
+    for (const auto &node : graphNodes) {
+        if (node->getType() == Type::FullyConnected) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/graph.cpp:1453]:
```
if (auto *fcNodePtr = dynamic_pointer_cast<...>(node)) {
    fcNodePtr->setMinSparseRate(minSparseRate);
}
```

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp [hunk 7]
```diff
@@ -575,6 +586,15 @@ void FullyConnected::createDescriptor(const std::vector<MemoryDescPtr> &inputDes
                              MemoryDescUtils::convertToDnnlMemoryDesc(outDesc)->getDnnlDesc());
 }
 
+void FullyConnected::createPrimitive() {
+    Node::createPrimitive();
+    // WA: We update implType to know whether weights decompression was used inside the kernel
+    NodeDesc *selected_pd = getSelectedPrimitiveDescriptor();
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/fullyconnected.cpp:592]: Impl type should be update inside prepareParams() function to satisfy dynamism flow.

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp [hunk 8]
```diff
@@ -668,6 +688,73 @@ InferenceEngine::Precision FullyConnected::getRuntimePrecision() const {
     return getMaxPrecision(inputPrecisions);
 }
 
+bool FullyConnected::canUseSparseWeights() {
+    if (useSparseWeights == false) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/fullyconnected.cpp:692]: I believe we don't need this condition anymore given EXPERIMENTAL_FC_SPARSE_WEI_DECOMPRESSION_ENABLE 
 will be deleted

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp [hunk 9]
```diff
@@ -668,6 +688,73 @@ InferenceEngine::Precision FullyConnected::getRuntimePrecision() const {
...
+    if (!impl::cpu::x64::mayiuse(impl::cpu::x64::avx512_core_amx))
+        return false;
+
+    if (isDynamicNode()) {
+        return false;
+    }
+
+    auto inputDataType = DnnlExtensionUtils::IEPrecisionToDataType(getOriginalInputPrecisionAtPort(DATA_ID));
+    auto weightsDataType = DnnlExtensionUtils::IEPrecisionToDataType(getOriginalInputPrecisionAtPort(WEIGHTS_ID));
+    if (!one_of(inputDataType , memory::data_type::u8, memory::data_type::s8) || weightsDataType != memory::data_type::s8) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/fullyconnected.cpp:705]: Please use ov::Precision. Not necessary to convert to DNNL representation.

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp [hunk 10]
```diff
@@ -668,6 +688,73 @@ InferenceEngine::Precision FullyConnected::getRuntimePrecision() const {
...
+        return false;
+    }
+
+    // [todo] delete
+    printSparseRates = false; // useSparseWeights;
+
+    // calculate sparse rate
+    const auto constNode = std::dynamic_pointer_cast<Input>(getParentEdgeAt(WEIGHTS_ID)->getParent());
+    if (!constNode) {
+        IE_THROW() << "Cannot cast " << WEIGHTS_ID << " input to Input node for " << getName() << ".";
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/fullyconnected.cpp:715]: You might have some layers between Constant node and FC (Reshape for example). This case shouldn't fail the execution. Just disable sparse feature if you cannot access weights tensor.

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/fullyconnected_sparse.cpp [hunk 11]
```diff
@@ -0,0 +1,329 @@
+// Copyright (C) 2018-2022 Intel Corporation
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/tests/functional/subgraph_tests/src/fullyconnected_sparse.cpp:1]: I am convinced that it is better to extend existing CPU single layer test instead of separate one.
Please extract the test into separate PR to speedup merge of production functionality.

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp [hunk 12]
```diff
@@ -668,6 +688,73 @@ InferenceEngine::Precision FullyConnected::getRuntimePrecision() const {
...
+    if (printSparseRates) {
+        std::cout << getName() << ", weightsData.size() = " << elementsCount << ", zerosCounts = "
+            << zerosCounts << ", nnzCount = " << nnzCount << std::endl;
+    }
+
+    weiSparseRate = static_cast<float>(zerosCounts) / static_cast<float>(elementsCount);
+
+    // minSparseRate == 1 means that user did not specify min sparse rate and minSparseRate will be set automatically
+    // todo: [av] maybe change to -1?
+    if (minSparseRate == 1.f) {
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/src/nodes/fullyconnected.cpp:740]: Now the behavior should be adjusted. `minSparseRate == 1.0f` - default value that means sparsity is switched off. In that case we should keep load time of current path: canUseSparseWeights should return false in very beginning,

**Review by dmitry-gorokhov — COMMENTED** [review]
OV part

File: src/bindings/python/src/pyopenvino/core/properties/properties.cpp (hunk 1 above)
Comment (antonvor) [reply src/bindings/python/src/pyopenvino/core/properties/properties.cpp]: done

File: src/tests/ie_test_utils/common_test_utils/ov_tensor_utils.cpp (hunk 2 above)
Comment (antonvor) [reply src/tests/ie_test_utils/common_test_utils/ov_tensor_utils.cpp]: done

File: src/plugins/intel_cpu/src/nodes/fullyconnected.h (hunk 4 above)
Comment (antonvor) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.h]: done

File: src/plugins/intel_cpu/src/graph.cpp (hunk 5 above)
Comment (antonvor) [reply src/plugins/intel_cpu/src/graph.cpp]: changed, thanks

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp (hunk 3 above)
Comment (antonvor) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.cpp]: done, thanks

File: src/plugins/intel_cpu/src/graph.cpp (hunk 6 above)
Comment (antonvor) [reply src/plugins/intel_cpu/src/graph.cpp]: changed, thanks

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp (hunk 8 above)
Comment (antonvor) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.cpp]: removed, thanks

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp (hunk 9 above)
Comment (antonvor) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.cpp]: changed, thanks

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp (hunk 10 above)
Comment (antonvor) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.cpp]: disabled, thanks

File: src/plugins/intel_cpu/src/nodes/fullyconnected.cpp (hunk 12 above)
Comment (antonvor) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.cpp]: changed, thanks

File: src/plugins/intel_cpu/tests/functional/subgraph_tests/src/fullyconnected_sparse.cpp (hunk 11 above)
Comment (antonvor) [reply src/plugins/intel_cpu/tests/functional/subgraph_tests/src/fullyconnected_sparse.cpp]: will be done in separate PR, thanks

**Review by dmitry-gorokhov — APPROVED** [review]
In general lgmt.
We need to follow up in separate PRs:
1. Fix remaining comments and and corresponding tests.
2. Update user documentation

File: src/plugins/intel_cpu/src/nodes/fullyconnected.h [hunk 13]
```diff
@@ -42,6 +42,7 @@ class FullyConnected : public Node {
 
     void initSupportedPrimitiveDescriptors() override;
     void initOptimalPrimitiveDescriptor() override;
+    // void createPrimitive() override;
```
[the remark this answers is not in the document]
Comment (antonvor) [reply src/plugins/intel_cpu/src/nodes/fullyconnected.h:45]: removed here https://github.com/openvinotoolkit/openvino/pull/14901

File: src/inference/include/openvino/runtime/intel_cpu/properties.hpp [hunk 14]
```diff
@@ -47,5 +47,7 @@ namespace intel_cpu {
  */
 static constexpr Property<bool> denormals_optimization{"CPU_DENORMALS_OPTIMIZATION"};
 
+static constexpr Property<float> sparse_weights_decompression_rate{"SPARSE_WEIGHTS_DECOMPRESSION_RATE"};
```
[the remark this answers is not in the document]
Comment (antonvor) [reply src/inference/include/openvino/runtime/intel_cpu/properties.hpp:50]: 1. added `CPU_` prefix here https://github.com/openvinotoolkit/openvino/pull/14901
2. will be done here https://github.com/openvinotoolkit/openvino/pull/14901
